### PR TITLE
Fix/close down blueberry endpoints

### PR DESCRIPTION
**Pull request by mayorcoded**
# Description

This PR implements the recommended fixes from the audit report for blocking all entry points into the blueberry contracts.

## Type of change

- [x] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [ ] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

**Review by Gornutz — APPROVED** [review]
- [x] Gornutz
- [ ] @0xSpraggins 
- [ ] @cuthalion0x

**Review by 0xSpraggins — CHANGES_REQUESTED** [review]
Hey @mayorcoded thanks for submitting this PR. Just a few formatting changes requested. I also think it would be beneficial to put permissions on all the `mint` and `burn` functions on the wrappers to make sure that only the associated spells is able to call those functions.

File: contracts/money-market/BToken.sol [hunk 1]
```diff
@@ -1302,6 +1302,18 @@ contract BToken is BTokenInterface, Exponential, TokenErrorReporter {
...
 
+    /**
+     * @notice allows the admin update the softVault address
+     * @dev Admin function to set the soft vault address
+     * @param _softVault the address of the SoftVault this contract is connected to
+     */
+    function _setSoftVault(address _softVault)
+    public
+    {
+        require(msg.sender == admin, "admin only");
```
[the remark this answers is not in the document]
Comment (mayorcoded) [reply contracts/money-market/BToken.sol:1313]: Thanks for this observation, I have added the changes

File: contracts/money-market/BToken.sol [hunk 2]
```diff
@@ -1302,6 +1302,22 @@ contract BToken is BTokenInterface, Exponential, TokenErrorReporter {
         return uint256(Error.NO_ERROR);
     }
 
+    /**
+     * @notice allows the admin update the softVault address
+     * @dev Admin function to set the soft vault address
+     * @param _softVault the address of the SoftVault this contract is connected to
+     */
+    function _setSoftVault(address _softVault) public returns(uint256){
```
Comment (0xSpraggins) [inline contracts/money-market/BToken.sol:1310]:
```suggestion
    function _setSoftVault(address _softVault) public returns (uint256) {
```
Just add spacing here

**Review by 0xSpraggins — APPROVED** [review]
Just some spacing changes but everything looks good and I'll approve. Ready for @cuthalion0x's review

**Review by cuthalion0x — COMMENTED** [review]
I feel that this PR should also introduce tests. I'm not very familiar with the test structure here, but to verify all of the new access control, we should test:

1. A user can call `SoftVault.deposit()` - this also tells us that the SoftVault can call `BToken.mint()`. This is probably already tested.
2. A user cannot call `BToken.mint()`. **This is not yet tested.**
3. A user can open a position within a Spell - this also tells us that the Bank can call `BToken.borrow()`. This is probably already tested for all Spells.
4. A user cannot call `BToken.borrow()`. **This is not yet tested.**

This is especially important for borrows. The access control for a mint is self-explanatory because we can see the requirement right at the start of execution. But for a borrow, we've introduced code into the Comptroller to enforce a constraint within the BToken. This relationship should be tested for all types of BTokens that we intend to deploy.

The remaining comments are just style fixes.

File: test/money-market/bToken.test.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      await expect(bWETH.mint(amount)).to.be.revertedWith('caller should be softvault');
+    });
+
+    it('should not revert when caller is soft vault', async () => {
+      const amount = ethers.BigNumber.from(ethers.utils.randomBytes(32));
+      await bTokenAdmin._setSoftVault(bUSDC.address, softVault.address);
+      await bTokenAdmin._setSoftVault(bWETH.address, softVault.address);
+
+      await expect(bUSDC.connect(softVault).mint(amount)).to.not.be.revertedWith('caller should be softvault');
+      await expect(bWETH.connect(softVault).mint(amount)).to.not.be.revertedWith('caller should be softvault');
```
Comment (cuthalion0x) [inline test/money-market/bToken.test.ts:49]: Personally, for a success test, I'd prefer checking the return value from the mint function. It should be zero on success.

File: test/money-market/bToken.test.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      await expect(bWETH.borrow(amount)).to.be.revertedWith('only bank can borrow');
+    });
+
+    it('should not revert when caller is bank', async () => {
+      const amount = ethers.BigNumber.from(ethers.utils.randomBytes(32));
+      await comptroller._setCreditLimit(bank.address, bUSDC.address, amount);
+      await comptroller._setCreditLimit(bank.address, bWETH.address, amount);
+
+      await expect(bUSDC.connect(bank).borrow(amount)).to.not.be.revertedWith('only bank can borrow');
+      await expect(bWETH.connect(bank).borrow(amount)).to.not.be.revertedWith('only bank can borrow');
```
Comment (cuthalion0x) [inline test/money-market/bToken.test.ts:66]: Same here, we expect borrow to return zero on success.

File: test/money-market/bToken.test.ts (hunk 3 above)
Comment (mayorcoded) [reply test/money-market/bToken.test.ts:57]: The tests have been updated

File: test/money-market/bToken.test.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      await expect(bUSDC.connect(softVault).mint(amount)).to.not.be.revertedWith('caller should be softvault');
+      await expect(bWETH.connect(softVault).mint(amount)).to.not.be.revertedWith('caller should be softvault');
+
+      await usdc.connect(softVault).approve(bUSDC.address, amount);
+      let returnValue = await bUSDC.connect(softVault).callStatic.mint(amount);
+      expect(returnValue).to.equal(0);
+
+      await weth.connect(softVault).approve(bWETH.address, amount);
+      returnValue = await bWETH.connect(softVault).callStatic.mint(amount);
+      expect(returnValue).to.equal(0);
```
Comment (0xSpraggins) [inline test/money-market/bToken.test.ts:65]:
```suggestion
      let success = await bUSDC.connect(softVault).callStatic.mint(amount);
      expect(success).to.equal(0);

      await weth.connect(softVault).approve(bWETH.address, amount);
      success = await bWETH.connect(softVault).callStatic.mint(amount);
      expect(success).to.equal(0);
```

Rename return value to improve clarity

File: test/money-market/bToken.test.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      await bUSDC.connect(softVault).mint(amount);
+
+      let returnValue = await bUSDC.connect(bank).callStatic.borrow(amount.div(2));
+      expect(returnValue).to.equal(0);
+
+      await weth.connect(softVault).approve(bWETH.address, amount);
+      await bWETH.connect(softVault).mint(amount);
+
+      returnValue = await bWETH.connect(bank).callStatic.borrow(amount.div(2));
+      expect(returnValue).to.equal(0);
```
Comment (0xSpraggins) [inline test/money-market/bToken.test.ts:96]:
```suggestion
      let success = await bUSDC.connect(bank).callStatic.borrow(amount.div(2));
      expect(success).to.equal(0);

      await weth.connect(softVault).approve(bWETH.address, amount);
      await bWETH.connect(softVault).mint(amount);

      success = await bWETH.connect(bank).callStatic.borrow(amount.div(2));
      expect(success).to.equal(0);
```

Rename return value to improve clarity

**Review by cuthalion0x — APPROVED** [review]

**Review by Gornutz — APPROVED** [review]

**Review by 0xSpraggins — APPROVED** [review]

**Pull request merged**